### PR TITLE
logging mode switch + automatic log file transfer + speed based logging

### PR DESCRIPTION
**Pull request by patcmorneau**

### CODE REVIEW

File: src/workspace/src/hydroball_data_websocket/include/hydroball_data_websocket/hydroball_data_websocket.h [hunk 1]
```diff
@@ -54,9 +54,10 @@ class TelemetryServer {
 
     void on_message(connection_hdl hdl, server::message_ptr msg) {
 	rapidjson::Document document;
-
         if(document.Parse(msg->get_payload().c_str()).HasParseError()){
         	//Not valid JSON
+        	ROS_ERROR("invalid json");
+        	std::cerr<<msg->get_payload().c_str()<<"\n";
```
Comment (glabmoris) [inline src/workspace/src/hydroball_data_websocket/include/hydroball_data_websocket/hydroball_data_websocket.h:60]: You can put line 59/60 in the ROS_ERROR call. ROS_ERROR is a C-style printf() wrapper. You can use printf flags to format error messages into it. Ex:

ROS_ERROR("Apples, oranges and %s", "bananas");

Or use ROS_ERROR_STREAM for a c++ stream syntax:

ROS_ERROR_STREAM("Apples, oranges and " << "bananas");

File: src/workspace/src/logger_text/src/logger_text.cpp [hunk 2]
```diff
@@ -5,14 +5,37 @@
...
+			configurationClient = node.serviceClient<setting_msg::ConfigurationService>("get_configuration");
+			ROS_INFO("getting speed threshold configuration...");
+			getSpeedThresholdConfig();
 }
 
 Writer::~Writer(){
 	finalize();
 }
 
+void Writer::getSpeedThresholdConfig(){
```
Comment (glabmoris) [inline src/workspace/src/logger_text/src/logger_text.cpp:19]: The intent is good but the name could clash with a common interpretation. Don't use an accessor syntax (getXXX) for a void function. Use something more meaningful such as updateSpeedThreshold()

File: src/workspace/src/logger_text/src/logger_text.cpp [hunk 3]
```diff
@@ -5,14 +5,37 @@
...
+	setting_msg::ConfigurationService srv;
+
+    srv.request.key = "speedThresholdKmh";
+
+    if(configurationClient.call(srv)){
+    	std::string speedThresKmh = srv.response.value;
+    	ROS_INFO_STREAM("speed threshold from config file : "<< speedThresKmh);
+        speedThresholdKmh = stod(speedThresKmh);
```
Comment (glabmoris) [inline src/workspace/src/logger_text/src/logger_text.cpp:27]: stod will throw an exception with an invalid string, crashing the whole thing. Add error checking and/or proper exception handling.

File: src/workspace/src/logger_text/src/logger_text.cpp [hunk 4]
```diff
@@ -125,38 +148,54 @@ void Writer::imuCallback(const sensor_msgs::Imu& imu){
...
-		kmh_Speed_list.push_back(current_speed);
-		average_speed = std::accumulate(kmh_Speed_list.begin(), kmh_Speed_list.end(), average_speed) / 120;
-		logger_service::GetLoggingStatus::Request request;
-		logger_service::GetLoggingStatus::Response response;
+	if(automaticMode){
+		speedThresholdKmh = getSpeedThreshold();
+
+		if(speedThresholdKmh < 0 && speedThresholdKmh > 100){
+			speedThresholdKmh = defaultSpeedThreshold;
+			std::string speed = std::to_string(speedThresholdKmh);
```
Comment (glabmoris) [inline src/workspace/src/logger_text/src/logger_text.cpp:159]: No need to convert to std::String to use ROS_ERROR_STREAM

File: src/workspace/src/logger_text/src/logger_text.cpp [hunk 5]
```diff
@@ -125,38 +148,54 @@ void Writer::imuCallback(const sensor_msgs::Imu& imu){
...
+			ROS_ERROR_STREAM("invalid speed threshold, defaulting to "<<speed<<" Kmh");
+		}
+		else if(speedThresholdKmh == 0.0){
+			speedThresholdKmh = defaultSpeedThreshold;
+			std::string speed = std::to_string(speedThresholdKmh);
```
Comment (glabmoris) [inline src/workspace/src/logger_text/src/logger_text.cpp:164]: No need to convert to std::String to use ROS_ERROR_STREAM

PS: Redundant logic

**Review by glabmoris — CHANGES_REQUESTED** [review]

File: src/workspace/src/logger_text/src/logger_text.cpp (hunk 2 above)
Comment (patcmorneau) [reply src/workspace/src/logger_text/src/logger_text.cpp]: fix

File: src/workspace/src/logger_text/src/logger_text.cpp (hunk 3 above)
Comment (patcmorneau) [reply src/workspace/src/logger_text/src/logger_text.cpp]: fix

File: src/workspace/src/logger_text/src/logger_text.cpp (hunk 4 above)
Comment (patcmorneau) [reply src/workspace/src/logger_text/src/logger_text.cpp:163]: fix

File: src/workspace/src/logger_text/src/logger_text.cpp (hunk 5 above)
Comment (patcmorneau) [reply src/workspace/src/logger_text/src/logger_text.cpp]: removed redundant logic

**Comment by patcmorneau** [comment]
the mutex is problematic in:
```
bool Writer::toggleLogging(logger_service::ToggleLogging::Request & request,logger_service::ToggleLogging::Response & response){
	mtx.lock();

	if(bootstrappedGnssTime){
		if(!loggerEnabled && request.loggingEnabled){
			//Enabling logging, init logfiles
			loggerEnabled=true;
			init();
		}
		else if(loggerEnabled && !request.loggingEnabled){
			//shutting down logging, finalize logfiles
			loggerEnabled=false;
			finalize();
		}

		mtx.unlock();

		response.loggingStatus=loggerEnabled;
		return true;
	}

```

**Comment by patcmorneau** [comment]
first byte should be 0xb5 and got :0xffffffb5
not quite sure where those ff bytes are coming from but once that will be figured out, will just have to calculate the appropriate checksum bytes and zf9p unit test will be good to go

**Pull request merged**